### PR TITLE
Some minor cleanups to Move IR

### PR DESCRIPTION
**Pull request by sunshowers**
Prepare for the addition of bytecode functions to the IR.

### CODE REVIEW

**Comment by vgao1996** [comment]
Looks good to me. The failed test is likely flaky. I've seen it timing out a few times myself.

File: language/move-ir/types/src/ast.rs [hunk 1]
```diff
@@ -149,13 +152,16 @@ pub type TypeVar = Spanned<TypeVar_>;
 // TODO: This enum is completely equivalent to vm::file_format::Kind.
 //       Should we just use vm::file_format::Kind or replace both with a common one?
 /// The kind of a type. Analogous to `vm::file_format::Kind`.
-#[derive(Debug, PartialEq, Eq, Clone)]
+#[derive(Debug, Display, PartialEq, Eq, Clone)]
 pub enum Kind {
     /// Represents the super set of all types.
+    #[display(fmt = "all")]
```
Comment (tnowacki) [inline language/move-ir/types/src/ast.rs]: I removed this a while back because I felt it hurt the readability of the AST. Thoughts on reverting?

Comment (sunshowers) [reply language/move-ir/types/src/ast.rs]: Do you mean the declarative annotations? Personally I strongly prefer them to a separate impl because it's all in one place.

Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: What does all in one place mean here?

Comment (sunshowers) [reply language/move-ir/types/src/ast.rs]: The declaration and the Display annotation live together. I think of the Display impl as providing a lot of valuable information to understand the code — I looked at them several times while figuring out how the AST IR worked. But I might be weird 🤷

Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: My main worry is some of them get quite verbose. If they were all one line, I would consider using them to replace the syntax/grammar comments

But as it stands, I'd prefer to keep them separate

Comment (sunshowers) [reply language/move-ir/types/src/ast.rs]: Well, I definitely prefer to keep them together — I think most of them are one line.

Comment (sunshowers) [reply language/move-ir/types/src/ast.rs]: FWIW it wouldn't be a good idea to replace the docs either — cargo doc is a super valuable way to look at these giant AST enums.

Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: > Well, I definitely prefer to keep them together 

I just don't see much benefit from using the derivation annotation. I made the switch to remove them as I wanted to fully separate the AST from any impls, as I found it made it easier to grok the AST. I'd like to keep that separation in place

Comment (sunshowers) [reply language/move-ir/types/src/ast.rs]: I understand that, but my view is completely the opposite. I find it unnecessarily complex to have the AST be separate from any impls, and I found it to be much harder to grok the AST with the separate impls. I strongly prefer making this change.

Out differences might be irreconcilable. How do you wish to proceed?

File: language/move-ir/types/src/ast.rs [hunk 2]
```diff
@@ -164,35 +170,52 @@ pub enum Kind {
...
+    #[display(fmt = "&{}{}", "if *_0 { \"mut \" } else { \"\" }", _1)]
     Reference(bool, Box<Type>),
     /// A type parameter
+    #[display(fmt = "{}", _0)]
     TypeParameter(TypeVar_),
 }
 
+/// A list of actual types.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct TypeActuals {
```
Comment (bob-wilson) [inline language/move-ir/types/src/ast.rs]: What is the reason to add this separate struct? It doesn't seem like it is anything more than just a vector of types. I suppose you might get a marginal improvement in type safety, but it's at the expense of transparency and simplicity. Is there some other motivation that I'm missing?

File: language/move-ir/types/src/ast.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: All of the ASTs in the source language follow this same pattern. So for now, differences aside, I think the best thing to do is to be consistent

File: language/move-ir/types/src/ast.rs (hunk 2 above)
Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: Didn't see this (thought it was just a soft type alias based on the usage). I'm going to have to agree with Bob that I don't think this change buys much

Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: In other words, I think a soft type alias would do much better here

Comment (sunshowers) [reply language/move-ir/types/src/ast.rs]: I disagree -- a struct is much better here. From the commit message:

> This has two purposes:

> 1. Makes it easy to distinguish between type actuals and the Vec<Type> used as the return type from functions.
> 2. Allows for a cleaner `Display` impl.

File: language/move-ir/types/src/ast.rs (hunk 1 above)
Comment (sunshowers) [reply language/move-ir/types/src/ast.rs]: I'm happy to make the same change to the source language ASTs as well, as much as it makes sense.

Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: I don't really value the change for the source language ASTs. And I'm really the only person using that code at the moment, so I'd like to keep it the way that let's me move as fast as possible

File: language/move-ir/types/src/ast.rs (hunk 2 above)
Comment (bob-wilson) [reply language/move-ir/types/src/ast.rs]: Ah, I had missed the explanation in the commit message.... I don't know much about Rust's Display thingy, so I'll defer to the rest of you on that. My personal inclination would be to keep it simple.

File: language/move-ir/types/src/ast.rs (hunk 1 above)
Comment (sunshowers) [reply language/move-ir/types/src/ast.rs]: Side note: I'm trying my best to understand your point of view, but I'm very surprised at the pushback to this. To me this is just plainly better. It is more declarative, it minimizes repeating oneself, and it really shines for larger enums. My specific motivation is that I would like to use this pattern for the IR bytecode, and I would be quite unhappy if I had to write out a `Display` impl by hand.

File: language/move-ir/types/src/ast.rs (hunk 2 above)
Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: I think (1) is also achieved by a soft type alias. And often times in the compiler you actually WANT the two to be interchangeable (i.e. you want to easily have code that works for both without jumping through a trait)

I agree with (2) that the implementation is cleaner, but I don't think the benefit outweighs the cost

File: language/move-ir/types/src/ast.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: Keeping all of impls of the AST nodes separate from the declaration makes it much easier to grok translation functions. You can easily side-by-side see the transformations that much occur. Adding in other data or functionality forces me to try to figure out what information I can discard mentally before doing the comparison.

Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: The same sort of principle occurs when programming against the AST itself. I often have two editors open, one of the AST used as a map of the landscape, and the other as the transformation/pass that I am writing. The cleaner the AST, the better the map

File: language/move-ir/types/src/ast.rs (hunk 2 above)
Comment (sunshowers) [reply language/move-ir/types/src/ast.rs]: > (i.e. you want to easily have code that works for both without jumping through a trait)

That trait is `IntoIterator`, and there was just one place where it was necessary to do that. It was quite easy to do in any case.

> I agree with (2) that the implementation is cleaner, but I don't think the benefit outweighs the cost

I do...

Comment (tnowacki) [reply language/move-ir/types/src/ast.rs]: >> I agree with (2) that the implementation is cleaner, but I don't think the benefit outweighs the cost

> I do...

I'm not really clear on the benefit here other than being able to declare additional impls easily against this type. 

Often the compiler, you want less constructs in your AST, not more. The reason being is you want as writing passes you want similar constructs to have the same code, as it forces you into a space of being more principled with your judgements/rules/translations, and forces you to think if you really need an ad hoc pass. In this case, I want the type actuals to be the same as any other vector of types, that is an intentional and desirable design choice. The reason being is that any vector/list of types should behave in the same way in their principles regardless of the usage. This change would make the default/easy option the ad hoc one, and would force you out of the default world of shared, principled passes. 

In short, this design choice to share `Vec<Types>` everywhere might seem lazy or unsafe in other software applications, but it is a specific and intentional choice for the compiler.

**Comment by sunshowers** [comment]
It seems like I've been unable to build consensus on this.
